src/test/AludelV3.t.sol

```diff
@@ -296,4 +270,359 @@ contract AludelV3Test is Test {
+        public
+    {
+        vm.assume(schedules > 0);
+        vm.assume(scheduleDuration > 0);
+
+        // I'm capping the amounts to avoid shares overflow
+        stakingAmount = bound(stakingAmount, 1, 1_000_000_000_000 ether);
+        rewardAmount = bound(rewardAmount, 1, 1_000_000_000_000 ether);
+
+        Crucible crucible = Utils.createCrucible(user, crucibleFactory);
+        Utils.fundMockToken(address(crucible), stakingToken, stakingAmount * schedules);
+
+        AludelV3.AludelData memory data = aludel.getAludelData();
+
+        Utils.stake(user, crucible, aludel, stakingToken, stakingAmount);
+
+        for (uint256 i = 0; i < schedules; i++) {
+            Utils.fundAludel(aludel, admin, rewardToken, rewardAmount, scheduleDuration);
+            vm.warp(block.timestamp + scheduleDuration);
+        }
+
+        uint256[] memory indices = new uint256[](1);
+        uint256[] memory amounts = new uint256[](1);
+        indices[0] = 0;
+        amounts[0] = stakingAmount;
+
+        Utils.unstake(user, crucible, aludel, stakingToken, indices, amounts);
+    }
+
+    function test_fund_without_approval(uint64 reward, uint32 duration) public {
+        vm.assume(reward > 0);
+        vm.assume(duration > 0);
+
+        Utils.fundMockToken(admin.addr(), rewardToken, uint256(reward));
+
+        vm.startPrank(admin.addr());
+
+        // before call fund we should approve the aludel to transfer the reward tokens
+
+        vm.expectRevert("TransferHelper::transferFrom: transferFrom failed");
+        aludel.fund(reward, duration);
+
+        assertEq(rewardToken.allowance(admin.addr(), address(aludel)), 0);
+    }
+
+    function test_fund_remove_expired_schedules(uint64 reward, uint32 duration) public {
+        vm.assume(reward > 0);
+        vm.assume(duration > 1);
+
+        Utils.fundMockToken(admin.addr(), rewardToken, uint256(reward) * 3);
+
+        vm.startPrank(admin.addr());
+
+        rewardToken.approve(address(aludel), uint256(reward) * 3);
+
+        IAludelV3.AludelData memory data = aludel.getAludelData();
+        // before fund there should be no schedules
+        assertEq(data.rewardSchedules.length, 0);
+
+        aludel.fund(reward, duration);
+
+        // after fund there should be one schedule
+        assertEq(aludel.getAludelData().rewardSchedules.length, 1);
+
+        // fund again right before the previous schedule expires
+        vm.warp(block.timestamp + duration - 1);
+        aludel.fund(reward, duration);
+
+        // so there should be two schedules
+        assertEq(aludel.getAludelData().rewardSchedules.length, 2);
+
+        // now the first schedule should expired
+        vm.warp(block.timestamp + 1);
+        aludel.fund(reward, duration);
+        // so there should be two schedules
+        assertEq(aludel.getAludelData().rewardSchedules.length, 2);
+    }
+}
+
+contract Given_A_New_Aludel_Is_Being_Funded is Test {
```

Comment:
This contract has a lot of code duplication with AludelV3Test

Please create a 3rd contract with the setup common to both and make them extend the setUp method and add whatever extra state variables are necessary.

Reply:
I agree with this, but I was hesitant to add it right away.
Would you agree to have a few more iterations of these test contracts before creating the common contract for all tests?

Reply:
I'm not saying we should have a common contract for *all* tests, that feels like premature optimization given the current state of affairs, I was talking about just this testfile.

If you look at both contracts side by side, you can see the repetition easily:
![image](https://user-images.githubusercontent.com/111788132/223143510-32ec9e43-8768-4c85-bcd3-632cad4f2689.png)

I don't like the idea of merging so much repeated code, but I wouldn't mind it in a draft PR